include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -44,6 +54,97 @@ class __radix_sort_reorder_peer_kernel;
 template <::std::uint32_t, bool, bool, typename... _Name>
 class __radix_sort_reorder_kernel;
 
+//------------------------------------------------------------------------
+// radix sort: bitwise order-preserving conversions to unsigned integrals
+//------------------------------------------------------------------------
+
+#if __cpluslplus >= 202002L && __has_include(<bit>)
+template <typename _Dst, typename _Src>
+using __dpl_bit_cast = std::bit_cast<_Dst, _Src>;
+
+#elif __has_builtin(__builtin_bit_cast)
+template <typename _Dst, typename _Src>
+_Dst
+__dpl_bit_cast(const _Src& __src)
+{
+    static_assert(sizeof(_Dst) == sizeof(_Src), "Bit conversion for types of different size");
+    return __builtin_bit_cast(_Dst, __src);
+}
+
+#else
+template <typename _Dst, typename _Src>
+_Dst
+__dpl_bit_cast(const _Src& __src)
+{
+    static_assert(sizeof(_Dst) == sizeof(_Src), "Bit conversion for types of different size");
+    _Dst __result;
+    std::memcpy(&__result, &__src, sizeof(_Dst));
+    return __result;
+}
+#endif
+
+template <bool __is_ascending>
+bool
+__order_preserving_cast(bool __val)
+{
+    if constexpr (__is_ascending)
+        return __val;
+    else
+        return !__val;
+}
+
+template <bool __is_ascending, typename _UInt, typename = __enable_if_t<::std::is_unsigned_v<_UInt>>>
+_UInt
+__order_preserving_cast(_UInt __val)
+{
+    if constexpr (__is_ascending)
+        return __val;
+    else
+        return ~__val; //bitwise inversion
+}
+
+template <bool __is_ascending, typename _Int,
+          typename = __enable_if_t<::std::is_integral_v<_Int> && ::std::is_signed_v<_Int>>>
+::std::make_unsigned_t<_Int>
+__order_preserving_cast(_Int __val)
+{
+    using _UInt = ::std::make_unsigned_t<_Int>;
+    // mask: 100..0 for ascending, 011..1 for descending
+    constexpr _UInt __mask = (__is_ascending) ? 1 << std::numeric_limits<_Int>::digits : ~_UInt(0) >> 1;
+    return __val ^ __mask;
+}
+
```

Comment:
I think readability and ease of understanding is almost always worth a few lines of redundant code. I will think though whether I can generalize it while keeping readability.

Reply:
I have taken another look. It appears to me that there is useful similarity, rather than duplication, between the overloads for floats and doubles. Surely it can be generalized, but essentially it would get back part of the infrastructure that has been removed, and even LOC would likely increase in the result.
Maybe if we add support for float16 etc. at some point. it will make sense to reconsider, but for now I will keep two overloads.